tsl/src/compression/recompress.c

```diff
@@ -0,0 +1,725 @@
+													Relation uncompressed_chunk_rel);
+static void update_current_segment(CompressedSegmentInfo *current_segment, TupleTableSlot *slot,
+								   int nsegmentby_cols);
+static void create_segmentby_scankeys(CompressionSettings *settings, Relation index_rel,
+									  Relation compressed_chunk_rel, ScanKeyData *index_scankeys);
+static void create_orderby_scankeys(CompressionSettings *settings, Relation index_rel,
+									Relation compressed_chunk_rel, ScanKeyData *orderby_scankeys);
+static void update_segmentby_scankeys(TupleTableSlot *uncompressed_slot,
+									  CompressedSegmentInfo *current_segment, int num_segmentby,
+									  ScanKey index_scankeys);
+static void update_orderby_scankeys(TupleTableSlot *uncompressed_slot,
+									CompressedSegmentInfo *current_segment, int num_segmentby,
+									int num_orderby, ScanKey orderby_scankeys);
+static enum Batch_match_result match_tuple_batch(TupleTableSlot *compressed_slot, int num_orderby,
+												 ScanKey orderby_scankeys, bool *nulls_first);
+static bool check_changed_group(CompressedSegmentInfo *current_segment, TupleTableSlot *slot,
+								int nsegmentby_cols);
+static void recompress_segment(Tuplesortstate *tuplesortstate, Relation compressed_chunk_rel,
+							   RowCompressor *row_compressor);
+/*
+ * Recompress an existing chunk by decompressing the batches
+ * that are affected by the addition of newer data. The existing
+ * compressed chunk will not be recreated but modified in place.
+ *
+ * 0 uncompressed_chunk_id REGCLASS
+ * 1 if_not_compressed BOOL = false
+ */
+Datum
+tsl_recompress_chunk_segmentwise(PG_FUNCTION_ARGS)
+{
+	Oid uncompressed_chunk_id = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	bool if_not_compressed = PG_ARGISNULL(1) ? true : PG_GETARG_BOOL(1);
+
+	ts_feature_flag_check(FEATURE_HYPERTABLE_COMPRESSION);
+	TS_PREVENT_FUNC_IF_READ_ONLY();
+	Chunk *chunk = ts_chunk_get_by_relid(uncompressed_chunk_id, true);
+
+	if (!ts_chunk_is_partial(chunk))
+	{
+		int elevel = if_not_compressed ? NOTICE : ERROR;
+		elog(elevel,
+			 "nothing to recompress in chunk %s.%s",
+			 NameStr(chunk->fd.schema_name),
+			 NameStr(chunk->fd.table_name));
+	}
+	else
+	{
+		if (!ts_guc_enable_segmentwise_recompression)
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("segmentwise recompression functionality disabled, "
+							"enable it by first setting "
+							"timescaledb.enable_segmentwise_recompression to on")));
+		}
+		uncompressed_chunk_id = recompress_chunk_segmentwise_impl(chunk);
+	}
+
+	PG_RETURN_OID(uncompressed_chunk_id);
+}
+
+Oid
+recompress_chunk_segmentwise_impl(Chunk *uncompressed_chunk)
+{
+	Oid uncompressed_chunk_id = uncompressed_chunk->table_id;
+
+	/*
+	 * only proceed if status in (3, 9, 11)
+	 * 1: compressed
+	 * 2: compressed_unordered
+	 * 4: frozen
+	 * 8: compressed_partial
+	 */
+	if (!ts_chunk_is_compressed(uncompressed_chunk) && ts_chunk_is_partial(uncompressed_chunk))
```

Comment:
This condition reads a little weird, how can chunk be uncompressed and partially compressed at the same time? Might be not a topic for this PR, but we might have to rename `ts_chunk_is_compressed` to something like `ts_chunk_is_fully_compressed`.

Reply:
Yeah, that check seems to check for two bitfields which should always accompany each other (compressed and partially compressed). Having partially compressed without the compressed bit should be an internal error.

Maybe `Ensure` is more suitable for this check?